File: python/collatz-conjecture/collatz_conjecture.py
"""
Calculate how many steps it takes for a positive integer number greater than zero to reach one 
following the steps i the collatz conjecture.
"""

def steps(number):
    """
    Calculate how many steps it takes for a positive integer number greater than zero to reach one 
    following the steps i the collatz conjecture.
    number: The number to start with.
    returns: The number of steps take to get to zero.
    """
    if number <= 0:
        raise ValueError('Only positive integers are allowed')
    num_steps = 0
    while number != 1:
        if number % 2 == 0:
            number /= 2
        else:
            number = number * 3 + 1
        num_steps += 1
    return num_steps

File: go/collatz-conjecture/collatz_conjecture.go
package collatzconjecture

import "errors"

/*
Collatz Conjecture - How many iterations does it take to reach 1 following the
conjecture for the given number (n).
@param n int - The number to start with
@returns int - Number of steps to reach 0
@errors - Error if n <= 0.
*/
func CollatzConjecture(n int) (int, error) {
	if n <= 0 {
		return 0, errors.New("Value must be a positive integer.")
	}
	steps := 0
	value := n
	for value != 1 {
		steps++
		if value%2 == 0 {
			value = value / 2
		} else {
			value = (value * 3) + 1
		}
	}
	return steps, nil
}
